core/src/main/java/hudson/security/TokenBasedRememberMeServices2.java

```diff
@@ -165,8 +183,42 @@ protected UserDetails processAutoLoginCookie(String[] cookieTokens, HttpServletR
         if (j.isDisableRememberMe()) {
             cancelCookie(request, response);
             throw new InvalidCookieException("rememberMe is disabled");
-        } else {
-            return super.processAutoLoginCookie(cookieTokens, request, response);
+        }
+        if (cookieTokens.length != 3) {
+            throw new InvalidCookieException(
+                    "Cookie token did not contain 3" + " tokens, but contained '" + Arrays.asList(cookieTokens) + "'");
+        }
+        long tokenExpiryTime = getTokenExpiryTime(cookieTokens);
+        if (isTokenExpired(tokenExpiryTime)) {
+            throw new InvalidCookieException("Cookie token[1] has expired (expired on '" + new Date(tokenExpiryTime)
+                    + "'; current time is '" + new Date() + "')");
+        }
+        // Check the user exists. Defer lookup until after expiry time checked, to
+        // possibly avoid expensive database call.
+        UserDetails userDetails = getUserDetailsService().loadUserByUsername(cookieTokens[0]);
+        Objects.requireNonNull(userDetails, "UserDetailsService " + getUserDetailsService()
+                + " returned null for username " + cookieTokens[0] + ". " + "This is an interface contract violation");
+        // Check signature of token matches remaining details. Must do this after user
+        // lookup, as we need the DAO-derived password. If efficiency was a major issue,
+        // just add in a UserCache implementation, but recall that this method is usually
+        // only called once per HttpSession - if the token is valid, it will cause
+        // SecurityContextHolder population, whilst if invalid, will cause the cookie to
+        // be cancelled.
+        String expectedTokenSignature = makeTokenSignature(tokenExpiryTime, userDetails.getUsername(),
+                userDetails.getPassword());
+        if (!equals(expectedTokenSignature, cookieTokens[2])) {
+            throw new InvalidCookieException("Cookie token[2] contained signature '" + cookieTokens[2]
+                    + "' but expected '" + expectedTokenSignature + "'");
+        }
+        return userDetails;
+    }
+
+    private long getTokenExpiryTime(String[] cookieTokens) {
+        try {
+            return Long.parseLong(cookieTokens[1]);
+        } catch (NumberFormatException nfe) {
+            throw new InvalidCookieException(
+                    "Cookie token[1] did not contain a valid number (contained '" + cookieTokens[1] + "')");
```

Comment:
Taken from https://github.com/spring-projects/spring-security/blob/0f64d4c091da2f0003c03b4940d713778fa8057e/web/src/main/java/org/springframework/security/web/authentication/rememberme/TokenBasedRememberMeServices.java#L123-L129 with a minor adaptation from `new Long` to `Long.parseLong` to fix SpotBugs errors.